### PR TITLE
Add Two Factor authentication

### PR DESCRIPTION
**Pull request by markstuart**
This PR encompasses the changes required to enforce 2fa for all users. It allows for users to reset their 2fa secret while logged in, and this same mechanism allows admins to reset the 2fa secret for a non-admin user (for example, if the user was to lose access to their Authenticator app).

Major changes:
- Login page now submits via ajax, allowing for 2 step login flow (username/password validation, then 2fa code submission). This means the user is not logged in fully until all checks have passed.
- New settings page for testing/resetting 2fa codes
- New db table for 2fa information, linked to user table
- New paster command for resetting a user's 2fa secret (useful for login flow in automated testing scenarios, but also as a backup for sysadmins to reset a users secret from the commandline)

Screenshot from 2020-02-28 10-10-54

Screenshot from 2020-02-28 10-12-33

### CODE REVIEW

File: ckanext/security/fanstatic/login_ajax.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      url: '/mfa_login',
+      method: 'POST',
+      data: form.serialize(),
+      success: function (loginState) {
+        hideError()
+
+        if (loginState.hasOwnProperty('mfaCodeValid')) {
+          handleMfaValidation(loginState.mfaCodeValid)
+          return
+        }
```
Comment (markstuart) [inline ckanext/security/fanstatic/login_ajax.js]: First form submit will validate the username and password only, so we show the 2fa form if it succeeds. The next submit has the username, password and 2fa code, and if that all succeeds we detach the form submit handler and do a standard form post, completing the login.

File: ckanext/security/fanstatic/login_ajax.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+      size: 250,
+      element: $('#qr-code-container')[0],
+      value: loginState.totpChallengerURI
+    })
+  }
+
+  var handleMfaValidation = function (validCode) {
+    if (validCode) {
+      getLoginForm().off('submit', validateLoginCredentials)
+      getLoginForm().submit()
```
Comment (matt-in-a-hat) [inline ckanext/security/fanstatic/login_ajax.js]: How does this result in different behaviour, such as being redirected somewhere, instead of spitting json back out at you?

File: ckanext/security/plugin.py [hunk 3]
```diff
@@ -1,26 +1,36 @@
...
 import ckan.logic.schema
+from ckanext.security.model import define_security_tables
+import logging
 
 from ckanext.security import schema
 
 
+
+log = logging.getLogger(__name__)
+userController = 'ckanext.security.controllers:SecureUserController'
```
Comment (matt-in-a-hat) [inline ckanext/security/plugin.py]: 1/10 No remaining reason this was moved up here?

File: ckanext/security/templates/security/configure_mfa.html [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  <section class="module module-narrow module-shallow">
+    <h2 class="module-heading"><i class="fa fa-medium fa-info-circle"></i> {{ _('Manage two factor authentication') }}</h2>
+    <div class="module-content">
+        {% if c.totp_challenger_uri %}
+        <p>
+            {% trans %}
+            Scan the image with the two factor authentication app on your phone.
+            {% endtrans %}
+        </p>
+        {% else %}
```
Comment (matt-in-a-hat) [inline ckanext/security/templates/security/configure_mfa.html]: 1/10 empty?

File: ckanext/security/templates/security/configure_mfa.html [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                    <canvas class="radius-lg padding-sm margin-b-sm border-solid" id="qr-code-container"></canvas>
+                </div>
+                <p>
+                    {% trans %}If you are not able to scan the QR code, you can manually enter this secret into your authenticator app:{% endtrans %} <code id="totp-secret">{{c.totp_secret}}</code>
+                </p>
+
+                {{ form.input('mfa', label=_("Enter the two factor authentication code from your app to test correct configuration"), id='field-mfa', type="text", value="", error=c.mfa_test_invalid, classes=["control-medium"]) }}
+                <button class="btn btn-primary" type="submit">{{_('Test')}}</button>
+                {% if c.mfa_test_valid %}
+                    <p>{{_('That\'s the correct code! Your authenticator app is correctly configured for future use.')}}</p>
```
Comment (matt-in-a-hat) [inline ckanext/security/templates/security/configure_mfa.html]: 😂 this sounds patronising to me "Wow, well done, you correctly typed in a number! Gold stars all around ⭐️"

File: ckanext/security/templates/security/configure_mfa.html [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+        {% set locale = h.dump_json({'content': _('Are you sure you want to generate a new two factor secret?')}) %}
+        <a class="btn btn-danger" href="{{new_mfa_challenge}}" data-module="confirm-action" data-module-i18n="{{ locale }}">{{_('Generate new two factor secret')}}</a>
+        <p>{{_('This should only be necessary if you have deleted the authenticator app or are replacing your phone.')}}</p>
+    {% else %}
+        <fieldset>
+            <legend>{{_('Add two factor authentication to your account')}}</legend>
+            <p>
+                {% trans %}
+                    Two factor authentication of your account will help protect your account from hackers. When logging in you will be required to provide a secret code generated from an authenticator app.
```
Comment (matt-in-a-hat) [inline ckanext/security/templates/security/configure_mfa.html]: I feel like the term "unauthorised access" is more professional than "hackers", but 🤷‍♂

File: ckanext/security/fanstatic/login_ajax.js (hunk 1 above)
Comment (markstuart) [reply ckanext/security/fanstatic/login_ajax.js]: Yes, that was the main reason... if they submit the first username/password without the 2fa code, the authentication mechanism will fail as it requires all 3 to succeed. The two step login was preferred rather than all in the same form, this seemed like a relatively simple way to provide error/validation feedback for username and password, and then the 2fa code afterwards.

File: ckanext/security/fanstatic/login_ajax.js (hunk 2 above)
Comment (markstuart) [reply ckanext/security/fanstatic/login_ajax.js]: Detaching the submit handler means that the subsequent "submit" call does the full form POST to `/login_generic` which is a non-ajax controller action.

File: ckanext/security/fanstatic/login_ajax.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+          handleMfaValidation(loginState.mfaCodeValid)
+          return
+        }
+
+        showMfaForm(loginState)
+      },
+      error: function (xhr) {
+        if (xhr.status === 403) {
+          showError('login')
+        }
```
[the remark this answers is not in the document]
Comment (markstuart) [reply ckanext/security/fanstatic/login_ajax.js]: Yes, I think you're right... this goes along with your comment about content-type changing response type. I actually doubt that it does, will do some investigation.

File: ckanext/security/fanstatic/qrcode_renderer.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
+'use strict';
+
+(function($) {
+    $(document).ready(function() {
+        var qrRenderTarget = document.getElementById("qr-code-container")
+        var totpUri = $("#totp-uri");
```
[the remark this answers is not in the document]
Comment (markstuart) [reply ckanext/security/fanstatic/qrcode_renderer.js]: I'll get that tidied up :+1:

File: ckanext/security/fanstatic/login_ajax.js (hunk 2 above)
Comment (matt-in-a-hat) [reply ckanext/security/fanstatic/login_ajax.js]: Ohh the form itself has a diff action URL, gotcha :+1:

File: ckanext/security/controllers.py [hunk 9]
```diff
@@ -81,40 +80,47 @@ def login(self):
...
-            return
 
-        identity = request.params
-        if not ('login' in identity and 'password' in identity):
-            tk.response.status_int = 422
-            return
+        try:
+            if request.method != 'POST':
+                tk.response.status_int = 405
+                return
```
Comment (matt-in-a-hat) [inline ckanext/security/controllers.py]: With these returns, is this still somewhere that could result in outputting a page after setting response type to json?
